### ISSUE TITLE
xAPI: correct response pattern

### ISSUE THREAD

**Issue by garemoko** [h5p/h5p-blanks#5]
The correct response pattern has two issues:

1. Hints are appended to the options. If this data is useful it should be included instead in a definition extension. 

2. Possible answers are delimited with a slash. instead they should be listed are multiple possible correct answers. So, for the example statement below:

```
"correctResponsesPattern": [
    "{case_matters=true}browser[,]Tin Can",
    "{case_matters=true}browser[,]Tin Can API",
    "{case_matters=true}web-browser[,]Tin Can",
    "{case_matters=true}web-browser[,]Tin Can API"
]
```

I appreciate this could be a lot of permutations where there are lots of blanks, but that's the spec as it is today and of course these permutations can be generated programmatically. 

Example statement from H5P wordpress plugin (updated today).

```
{
    "id": "12fed9ff-f26e-42ae-bb01-1e41f1526008",
    "actor": {
        "name": "admin",
        "mbox": "mailto:andrew.downes@watershedlrs.com",
        "objectType": "Agent"
    },
    "verb": {
        "id": "http://adlnet.gov/expapi/verbs/answered",
        "display": {
            "en-US": "answered"
        }
    },
    "result": {
        "score": {
            "scaled": 0,
            "raw": 0,
            "min": 0,
            "max": 2
        },
        "completion": true,
        "response": "web browser[,]tin can",
        "duration": "PT282.78S"
    },
    "context": {
        "contextActivities": {
            "category": [
                {
                    "id": "http://h5p.org/libraries/H5P.Blanks-1.4",
                    "objectType": "Activity"
                }
            ],
            "grouping": [
                {
                    "id": "http://localhost:8888/wordpress/h5p-test/",
                    "definition": {
                        "name": {
                            "en": " | H5P testWordPress sandbox"
                        },
                        "type": "http://activitystrea.ms/schema/1.0/page",
                        "moreInfo": "http://localhost:8888/wordpress/h5p-test/"
                    },
                    "objectType": "Activity"
                }
            ]
        }
    },
    "timestamp": "2016-01-18T18:06:35.886Z",
    "stored": "2016-01-18T18:06:35.886Z",
    "authority": {
        "name": "H5P",
        "account": {
            "homePage": "https://sandbox.watershedlrs.com",
            "name": "CDrGGtiYAEjUFr"
        },
        "objectType": "Agent"
    },
    "version": "1.0.0",
    "object": {
        "id": "http://localhost:8888/wordpress/wp-admin/admin-ajax.php?action=h5p_embed&id=9",
        "definition": {
            "extensions": {
                "http://h5p.org/x-api/h5p-local-content-id": 9
            },
            "name": {
                "en-US": "Fill in the blanks"
            },
            "description": {
                "en-US": "Fill in the missing words
H5P content may be edited using a __________.

\n
xAPI is also known as __________.

\n"
            },
            "type": "http://adlnet.gov/expapi/activities/cmi.interaction",
            "interactionType": "fill-in",
            "correctResponsesPattern": [
                "{case_matters=true}browser/web-browser:Something you use every day[,]Tin Can/Tin Can API:Contains Baked Beans"
            ]
        },
        "objectType": "Activity"
    }
}
```

**Comment by falcon-git** [comment]
Yeah, I think this was forgotten, I wasn't sure about adding all those permutations, but since you indicate it is the only way at the moment I'll add it, hopefully for the next release, and get rid of the hints as well obvously

**Comment by falcon-git** [comment]
This was fixed in:
8e364d0115b7a00171561c6752382c5e427b34cc
2f92059e078967234d99f5ebf5525a3e29747ef7
1bf7f97071ca17b457048d43a4fd7410024e35ae

**Comment by kienv** [comment]
Hi guys,

We found this becomes a real problem. We have a small activity with 14 fill-in-blank questions. Each question has 4 possible answers. current, code generate a 4^14 = 268435456 array for possible answers. this make browser got crashed and also DB issue. 

Could you please review it or do you guy already working on this?

**Comment by falcon-git** [comment]
@garemoko do you know if there are better ways of doing this now? If not I guess we should handle this by:

If the user answered incorrectly use the first alternative answer and the other ones in a custom property(extension), if the user correctly answered one of the alternative answers we use that answer and store alternatives in the same extension.

**Comment by garemoko** [comment]
Hi, I don't recall exactly how the blanks question works. Would it make sense to model each blank as a separate question with a separate xapi statement to reduce the number of permutations?

**Comment by falcon-git** [comment]
There might be multiple blanks in each sentence, i.e. There might be ______ _______ in each ________

I'm not sure if it would make sense to model it as multiple sentences. The real problem at least is how xAPI presents multiple alternative answers.

**Comment by garemoko** [comment]
We've also got to think about how the data will be consumed. xAPI is very unlikely to change at this point but even if it did, how would a reporting tool represent all those options? I think it would have to present them as though they were separate questions.

**Comment by kienv** [comment]
I think if we can keep them in order then i think it can reduce a lot of possible answers
eg . [first blank - possible answer 1,possible answer 2,possible answer 3,..],[second blank - possible answer 1,possible answer 2,possible answer 3,..],....

**Comment by ilukanets** [comment]
> Hi guys,
> 
> We found this becomes a real problem. We have a small activity with 14 fill-in-blank questions. Each question has 4 possible answers. current, the code generates a 4^14 = 268435456 rows array for possible answers. this makes browsers get crashed and also DB issue.
> 
> Could you please review it or do you guy already working on this?

Hello... 

We have the same issue reported by one of our customers. Are there some news on this?

**Comment by otacke** [comment]
@garemoko Hi! I was just made a community maintainer for H5P Group's repositories, and I am now trying to tackle all the open issues and pull requests.

I am sorry that this discussion seems to have died. But I'd like to summarize what happened since.

H5P Group (or Joubel back then) decided that generating all the permutations for exercises including potential optional answers in order to adhere to the xAPI specification was not feasible, as it stalled browsers when all permutations were computed and caused trouble further down the line when xAPI data was stored.

They switched to only listing the first correct answers of a question in the `correctResponsesPattern`, e.g. `["{case_matters=true}a[,]b"]` even though `a` and `b` might have equivalently correct alternatives. Note that hints are not attached to the answers.

In order to still be able to generate the full set of possible answers inside a gradebook or elsewhere where xAPI data end up, they introduced an extension `https://h5p.org/x-api/alternatives` that holds an array containing all the correct answer alternatives for all blanks, e.g. `[['a', 'A'], ['b', 'B']]`.

Hope that helps in any way. Please feel free to re-open the issue if you think that this needs to be solved differently and you could think of a feasible way.
